pennylane/tracker.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+This module contains a class for updating and recording information about device executions.
+"""
+
+
+class Tracker:
+    """This class stores information about device executions and allows users to interact with that
+    data upon each executions, even within parameter-shift gradients and optimization steps.
+
+    The information is stored in three class attributes: ``totals``, ``history``, and ``latest``.
+    Standard devices will track the number of executions, number of shots, and number of batch'
+    executions, but plugins may store additional information with no changes to this class.
+
+    Information is only stored when the class attribute ``tracking`` is set to ``True``. This
+    attribute can be toggled via a context manager and Python's ``with`` statement. Upon entering a
```

Comment:
Does `dev.tracking = True` work too?

Reply:
A `Tracker` instance is a device attribute. That instance controls tracking mode.

 We could create a getter and setter for `dev.tracking` that just points to `device.tracker.tracking`.

Reply:
My bad, meant `dev.tracker.tracking = True`.

Reply:
Yep. You can manaully control tracking mode.